File: simulation/engine/turn_order.py
from __future__ import annotations

import math
import random as _random_module
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from models.modifier import Modifier, STAT_SCALE
from models.enums import Stat, AiHeuristic
from engine.stats import calculate_stat

if TYPE_CHECKING:
    from models.card import Card

CT_THRESHOLD = 100 * STAT_SCALE  # 100000
HAND_SIZE = 5


@dataclass
class CombatEntity:
    id: str
    name: str
    base_stats: dict[Stat, int]
    active_modifiers: list[Modifier] = field(default_factory=list)
    ct: int = 0
    is_player: bool = True
    card_pool: list[str] = field(default_factory=list)
    ai_heuristic: AiHeuristic | None = None
    is_alive: bool = True
    current_energy: int = 0  # refreshed each turn start
    # Deck state (populated by initialize_deck at combat start)
    draw_pile: list[str] = field(default_factory=list)
    hand: list[str] = field(default_factory=list)
    discard_pile: list[str] = field(default_factory=list)


def get_current_stat(entity: CombatEntity, stat: Stat) -> int:
    return calculate_stat(entity.base_stats[stat], entity.active_modifiers, stat)


def tick_until_next_turn(entities: list[CombatEntity]) -> CombatEntity:
    """
    Advance CT for all living entities until one reaches CT_THRESHOLD.
    Tie-break: highest CT overflow, then highest Speed, then list position.
    """
    living = [e for e in entities if e.is_alive]
    if not living:
        raise ValueError("No living entities to tick")

    # Calculate ticks needed for each entity to reach threshold
    min_ticks = None
    for e in living:
        speed = get_current_stat(e, Stat.Speed)
        if speed <= 0:
            continue
        needed = max(0, CT_THRESHOLD - e.ct)
        ticks = math.ceil(needed / speed) if needed > 0 else 0
        if min_ticks is None or ticks < min_ticks:
            min_ticks = ticks

    if min_ticks is None:
        # All living entities have Speed <= 0 (e.g. everyone stunned simultaneously).
        # Fall back to list-order: pick the first living entity so combat can continue.
        living.sort(key=lambda e: entities.index(e))
        actor = living[0]
        actor.ct = 0
        return actor

    # Advance all living entities
    for e in living:
        speed = get_current_stat(e, Stat.Speed)
        e.ct += min_ticks * speed

    # Find ready entities (CT >= threshold)
    ready = [e for e in living if e.ct >= CT_THRESHOLD]
    if not ready:
        # Should not happen, but guard against floating-point edge cases.
        ready = living

    # Sort: highest overflow first, then highest speed, then list order
    ready.sort(key=lambda e: (-e.ct, -get_current_stat(e, Stat.Speed), entities.index(e)))

    actor = ready[0]
    actor.ct -= CT_THRESHOLD
    return actor


def process_turn_start(entity: CombatEntity, encounter_turn: int = 0) -> list[str]:
    """
    Called when entity's turn begins:
    1. Tick down durations on active_modifiers (duration > 0)
    2. Remove expired modifiers (duration == 0 after decrement)
    3. Check for death from effective HP <= 0
    4. Refresh Energy to calculated base
    Returns list of log messages.
    """
    logs = []

    # 1. Decrement durations for modifiers with duration > 0
    new_mods = []
    for m in entity.active_modifiers:
        if m.duration > 0:
            new_mod = m.model_copy(update={"duration": m.duration - 1})
            new_mods.append(new_mod)
        else:
            new_mods.append(m)
    entity.active_modifiers = new_mods

    # 2. Remove expired modifiers (duration == 0 after decrement; duration=-1 permanent stays)
    expired = [m for m in entity.active_modifiers if m.duration == 0]
    entity.active_modifiers = [m for m in entity.active_modifiers if m.duration != 0]
    if expired:
        logs.append(f"{entity.name}: {len(expired)} modifier(s) expired")

    # 3. Check death from DoT (effective HP <= 0)
    if entity.is_alive and get_current_stat(entity, Stat.HP) <= 0:
        entity.is_alive = False
        logs.append(f"{entity.name} has died")

    # 4. Refresh Energy
    entity.current_energy = get_current_stat(entity, Stat.Energy)

    return logs


# ---------------------------------------------------------------------------
# Deck system
# ---------------------------------------------------------------------------

def initialize_deck(
    entity: CombatEntity,
    cards_by_id: dict,
    rng: _random_module.Random,
) -> None:
    """Build draw pile from card_pool respecting deck_copies. Shuffle with seeded RNG."""
    if not cards_by_id:
        entity.draw_pile = []
        entity.hand = []
        entity.discard_pile = []
        return
    deck = []
    for card_id in entity.card_pool:
        card = cards_by_id.get(card_id)
        if card:
            copies = getattr(card, "deck_copies", 1)
            deck.extend([card_id] * copies)
    rng.shuffle(deck)
    entity.draw_pile = deck
    entity.hand = []
    entity.discard_pile = []


def draw_cards(entity: CombatEntity, count: int, rng: _random_module.Random) -> list[str]:
    """Draw `count` cards into hand. Reshuffles discard into draw pile when exhausted."""
    drawn = []
    for _ in range(count):
        if not entity.draw_pile:
            if not entity.discard_pile:
                break  # No cards left anywhere
            entity.draw_pile = list(entity.discard_pile)
            entity.discard_pile = []
            rng.shuffle(entity.draw_pile)
        if entity.draw_pile:
            card_id = entity.draw_pile.pop(0)
            entity.hand.append(card_id)
            drawn.append(card_id)
    return drawn


def discard_hand(entity: CombatEntity) -> None:
    """Move all cards from hand to discard pile."""
    entity.discard_pile.extend(entity.hand)
    entity.hand = []


def discard_card(entity: CombatEntity, card_id: str) -> None:
    """Move a specific card from hand to discard pile after playing it."""
    if card_id in entity.hand:
        entity.hand.remove(card_id)
        entity.discard_pile.append(card_id)

File: game/src/sim/turnOrder.ts
import { calculateStat } from "./stats";
import { CT_THRESHOLD, HAND_SIZE, Stat, type Card, type CombatEntity } from "./types";
import type { SeededRng } from "./rng";

export function createCombatEntity(partial: Partial<CombatEntity> & Pick<CombatEntity, "id" | "name" | "base_stats">): CombatEntity {
  return {
    active_modifiers: [],
    ct: 0,
    is_player: true,
    card_pool: [],
    ai_heuristic: null,
    is_alive: true,
    current_energy: 0,
    draw_pile: [],
    hand: [],
    discard_pile: [],
    ...partial
  };
}

export function getCurrentStat(entity: CombatEntity, stat: Stat): number {
  return calculateStat(entity.base_stats[stat], entity.active_modifiers, stat);
}

export function tickUntilNextTurn(entities: CombatEntity[]): CombatEntity {
  const living = entities.filter((entity) => entity.is_alive);
  if (living.length === 0) throw new Error("No living entities to tick");

  let minTicks: number | null = null;
  for (const entity of living) {
    const speed = getCurrentStat(entity, Stat.Speed);
    if (speed <= 0) continue;
    const needed = Math.max(0, CT_THRESHOLD - entity.ct);
    const ticks = needed > 0 ? Math.ceil(needed / speed) : 0;
    if (minTicks === null || ticks < minTicks) minTicks = ticks;
  }

  if (minTicks === null) {
    const actor = living.sort((a, b) => entities.indexOf(a) - entities.indexOf(b))[0];
    actor.ct = 0;
    return actor;
  }

  for (const entity of living) {
    entity.ct += minTicks * getCurrentStat(entity, Stat.Speed);
  }

  const ready = living.filter((entity) => entity.ct >= CT_THRESHOLD);
  const candidates = ready.length > 0 ? ready : living;
  candidates.sort((a, b) => {
    const ctDiff = b.ct - a.ct;
    if (ctDiff !== 0) return ctDiff;
    const speedDiff = getCurrentStat(b, Stat.Speed) - getCurrentStat(a, Stat.Speed);
    if (speedDiff !== 0) return speedDiff;
    return entities.indexOf(a) - entities.indexOf(b);
  });

  const actor = candidates[0];
  actor.ct = actor.ct >= CT_THRESHOLD ? actor.ct - CT_THRESHOLD : 0;
  return actor;
}

export function processTurnStart(entity: CombatEntity, _encounterTurn = 0): string[] {
  const logs: string[] = [];
  entity.active_modifiers = entity.active_modifiers.map((modifier) =>
    modifier.duration > 0 ? { ...modifier, duration: modifier.duration - 1 } : modifier
  );
  const expired = entity.active_modifiers.filter((modifier) => modifier.duration === 0);
  entity.active_modifiers = entity.active_modifiers.filter((modifier) => modifier.duration !== 0);
  if (expired.length > 0) logs.push(`${entity.name}: ${expired.length} modifier(s) expired`);
  if (entity.is_alive && getCurrentStat(entity, Stat.HP) <= 0) {
    entity.is_alive = false;
    logs.push(`${entity.name} has died`);
  }
  entity.current_energy = getCurrentStat(entity, Stat.Energy);
  return logs;
}

export function initializeDeck(entity: CombatEntity, cardsById: Record<string, Card>, rng: Pick<SeededRng, "shuffle">): void {
  const deck: string[] = [];
  for (const cardId of entity.card_pool) {
    const card = cardsById[cardId];
    if (card) {
      for (let i = 0; i < (card.deck_copies ?? 1); i += 1) deck.push(cardId);
    }
  }
  rng.shuffle(deck);
  entity.draw_pile = deck;
  entity.hand = [];
  entity.discard_pile = [];
}

export function drawCards(entity: CombatEntity, count: number, rng: Pick<SeededRng, "shuffle">): string[] {
  const drawn: string[] = [];
  for (let i = 0; i < count; i += 1) {
    if (entity.draw_pile.length === 0) {
      if (entity.discard_pile.length === 0) break;
      entity.draw_pile = [...entity.discard_pile];
      entity.discard_pile = [];
      rng.shuffle(entity.draw_pile);
    }
    const cardId = entity.draw_pile.shift();
    if (cardId === undefined) break;
    entity.hand.push(cardId);
    drawn.push(cardId);
  }
  return drawn;
}

export function discardHand(entity: CombatEntity): void {
  entity.discard_pile.push(...entity.hand);
  entity.hand = [];
}

export function discardCard(entity: CombatEntity, cardId: string): void {
  const index = entity.hand.indexOf(cardId);
  if (index >= 0) {
    entity.hand.splice(index, 1);
    entity.discard_pile.push(cardId);
  }
}

export { HAND_SIZE };
